### ISSUE TITLE
OperatorBench Plan

### ISSUE THREAD

**Issue by FindHao** [pytorch/pytorch#136168]
### 🚀 The feature, motivation and pitch

I'm working on benchmarking some custom triton kernels and different PyTorch operator implementations for inductor improvement.

Here are the goals
- [x] compare execution times of the custom triton kernels with compiled kernels
  -  ~~[initial PR](https://github.com/pytorch/pytorch/pull/136169)~~ https://github.com/pytorch/benchmark/pull/2485
  - https://github.com/pytorch/benchmark/pull/2507
  - https://github.com/pytorch/benchmark/pull/2508
  - https://github.com/pytorch/benchmark/pull/2509
  - https://github.com/pytorch/benchmark/pull/2517
- [x] compare native CUDA kernels (native PyTorch implementations) with compiled kernels (merge [the previous operatorbench](https://github.com/pytorch/pytorch/blob/main/benchmarks/dynamo/microbenchmarks/operatorbench.py)) 
  - https://github.com/pytorch/benchmark/pull/2495
- [x] add accuracy check
  - already implemented by tritonbench
- [x] add memory throughput, peak memory usage, and other performance metrics
  - https://github.com/pytorch/benchmark/pull/2490 
  - https://github.com/pytorch/benchmark/pull/2496
  - https://github.com/pytorch/benchmark/pull/2497
  - https://github.com/pytorch/benchmark/pull/2501
  - https://github.com/pytorch/benchmark/pull/2503

- [ ] add it to nightly testing and the PyTorch performance dashboard
  - https://github.com/pytorch/test-infra/pull/5839

cc @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @jansel

**Comment by shunting314** [comment]
It would be better to call out Triton: we want compare custom kernel (triton/cuda) perf with compiled kernel perf.

**Comment by FindHao** [comment]
> It would be better to call out Triton: we want compare custom kernel (triton/cuda) perf with compiled kernel perf.

Sounds good. updated the descriptions

**Comment by eellison** [comment]
Would be great to extend what @yanboliang has added instead of starting something new when possible. cc @yanboliang for thoughts.

Also, just to explicitly call out the okr:

```
Build a flywheel from hand-written custom kernels to compiler improvement that consists of a collection of custom kernels (incl. TritonBench, quantized custom kernels, etc)  w/ PyTorch-native reference implementation as a performance benchmark for Inductor
```

Potentially we could also benchmark against jax.

**Comment by FindHao** [comment]
> Would be great to extend what @yanboliang has added instead of starting something new when possible. cc @yanboliang for thoughts.
> 
> Also, just to explicitly call out the okr:
> 
> ```
> Build a flywheel from hand-written custom kernels to compiler improvement that consists of a collection of custom kernels (incl. TritonBench, quantized custom kernels, etc)  w/ PyTorch-native reference implementation as a performance benchmark for Inductor
> ```
> 
> Potentially we could also benchmark against jax.

Sounds good to me. I just went a quick through of gpt-fast. It is possible to reuse the operator definitions in gpt-fast. we can import gpt-fast and add some wrappers of the operations in gpt-fast. Just like the added FusedLinearCrossEntropyOperator. Jax is in the same case, like the added inductor.py for FusedLinearCrossEntropyOperator.

**Comment by xuzhao9** [comment]
We will be publishing tritonbench soon at https://github.com/pytorch/tritonbench, it would be great if we could reuse it as a library (similar to how PT2 Benchmark Runner is using torchbench)

**Comment by FindHao** [comment]
> We will be publishing tritonbench soon at https://github.com/pytorch/tritonbench, it would be great if we could reuse it as a library (similar to how PT2 Benchmark Runner is using torchbench).
> 
> Tritonbench will have CI tests and install scripts to setup all its dependencies (FBGEMM, HSTU, flash-attention, thunderkittens, etc (https://github.com/pytorch/benchmark/tree/main/submodules)) so that PyTorch developers don't need to worry about environment setup.
> 
> We are also working on Mosaic/Pallas support.

This is great! That's what I'd like to do, import tritonbench and reuse the operators defined there, and even some infrastructures.

**Comment by FindHao** [comment]
CC @jansel for thoughts.

**Comment by jansel** [comment]
We should be able to reuse existing PyTorch benchmark frameworks.  If we can't, I'd like to see a more thought out argument for why we need a new one.  We already have a lot of them and this feels like creating fragmentation.

**Comment by FindHao** [comment]
> We should be able to reuse existing PyTorch benchmark frameworks. If we can't, I'd like to see a more thought out argument for why we need a new one. We already have a lot of them and this feels like creating fragmentation.

I've discussed with @xuzhao9 and it's practical to achieve all our goals under tritonbench. I'm going to start migration now.

**Comment by FindHao** [comment]
Close this since all features are done except the last optional one.
